hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/records/ArrayPathsRecord.java

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright 2017 HugeGraph Authors
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership. The ASF
+ * licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.traversal.algorithm.records;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Random;
+import java.util.Stack;
+import java.util.function.Function;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.eclipse.collections.api.iterator.IntIterator;
+import org.eclipse.collections.impl.map.mutable.primitive.IntIntHashMap;
+
+import com.baidu.hugegraph.backend.id.Id;
+import com.baidu.hugegraph.perf.PerfUtil.Watched;
+import com.baidu.hugegraph.traversal.algorithm.HugeTraverser.Path;
+import com.baidu.hugegraph.traversal.algorithm.HugeTraverser.PathSet;
+import com.baidu.hugegraph.util.collection.ObjectIntMapping;
+import com.google.common.collect.Lists;
+
+public class ArrayPathsRecord implements Record {
+
+    private ObjectIntMapping idMapping;
+
+    private final Stack<Int2IntsMap> sourceLayers;
+    private final Stack<Int2IntsMap> targetLayers;
+
+    private Int2IntsMap currentLayer;
+    private IntIterator iterator;
+    private int current;
+    private boolean forward;
+
+    private int accessed;
+
+    public ArrayPathsRecord(Id sourceV, Id targetV) {
+
+        this.idMapping = new ObjectIntMapping();
+
+        int sourceCode = this.code(sourceV);
+        int targetCode = this.code(targetV);
+        Int2IntsMap firstSourceLayer = new Int2IntsMap();
+        Int2IntsMap firstTargetLayer = new Int2IntsMap();
+        firstSourceLayer.add(sourceCode, 0);
+        firstTargetLayer.add(targetCode, 0);
+        this.sourceLayers = new Stack<>();
+        this.targetLayers = new Stack<>();
+        this.sourceLayers.push(firstSourceLayer);
+        this.targetLayers.push(firstTargetLayer);
```

Comment:
prefer to split into two layers: fanout and source-target, and abstract source-target layer to the top of all record implement